amethyst_assets/src/storage.rs

```diff
@@ -30,9 +33,11 @@ impl Allocator {
 pub struct AssetStorage<A: Asset> {
     assets: VecStorage<A>,
     bitset: BitSet,
-    handles: Vec<Handle<A>>,
+    handles: Vec<WeakHandle<A>>,
```

Comment:
No need for using `Weak` inside `AssetStorage`
* Upgrading `Weak` has more overhead than cloning `Arc`. There is possibly infinite loop with CAS.
* You get nothing from `Weak` as you store simple `u32` and you don't have to drop it asap to free some resource. No memory get freed until you drop last `Weak`.
* And then you have to allocate new `Arc` after you finally drop `Weak`.

Using `WeakHandle` outside of `AssetStorage` may make some sense though.

Reply:
The biggest reason I used `WeakHandle` here is for consistency, since the `AssetStorage` does not in fact own a handle (logically). And for `reloads` it is indeed necessary.